prov/ucx/src/ucx_core.h

```diff
@@ -0,0 +1,392 @@
+		status = ucp_tag_msg_recv_nb(u_ep->worker, recv_buf, recv_cnt,
+					     recv_dt, cmsg->ucp_msg, cbf);
+		free(cmsg);
+	} else {
+		cbf = completion ?
+				ucx_recv_callback : ucx_recv_callback_no_compl;
+		status = ucp_tag_recv_nb(u_ep->worker, recv_buf, recv_cnt,
+					 recv_dt, msg->tag, ~msg->ignore, cbf);
+	}
+
+	if (UCS_PTR_IS_ERR(status)) {
+		FI_DBG(&ucx_prov,FI_LOG_CORE,
+		       "Recv operation returns error: %s",
+		       ucs_status_string(*(ucs_status_t*) status));
+		return UCX_TRANSLATE_ERRCODE(*(ucs_status_t *) status);
+	}
+
+	req = (struct ucx_request *) status;
+	req->cq = cq;
+	req->ep = u_ep;
+
+	if (msg->context) {
+		struct fi_context *ctx = ((struct fi_context *)(msg->context));
+		ctx->internal[0] = (void*)req;
+		ctx->internal[1] = NULL;
+	}
+
+	req->completion.tagged.op_context = msg->context;
+	req->completion.tagged.flags = FI_RECV |
+				       (mode == UCX_MSG ? FI_MSG : FI_TAGGED);
+	req->completion.tagged.buf = claim_discard ?
+					NULL : msg->msg_iov[0].iov_base;
+	req->completion.tagged.data = 0;
+	req->posted_size = posted_size;
+
+	/*
+	 * Nothing has arrived yet. The callback function will handle the
+	 * completion.
+	 */
+	if (req->type == UCX_FI_REQ_UNINITIALIZED) {
+		assert(!claim_discard);
+
+		req->type = UCX_FI_REQ_REGULAR;
+		req->completion.tagged.tag = msg->tag;
+		req->completion.tagged.len = msg->msg_iov[0].iov_len;
+		return FI_SUCCESS;
+	}
+
+	/*
+	 * Already done (matched an unexpected message). req->type is set
+	 * by the callback function.
+	 */
+	tc = &req->completion.tagged;
+	if (req->type == UCX_FI_REQ_UNEXPECTED_ERR &&
+	    !(claim_discard &&
+	      UCX_TRANSLATE_ERRCODE((int)req->status) == -FI_ETRUNC)) {
+		UCX_PUSH_ERROR_COMPLETION(cq, tc->op_context, tc->flags,
+					  (int)req->status,
+					  -UCX_TRANSLATE_ERRCODE((int)req->status),
+					  (tc->len - req->posted_size),
+					  ret, tc->tag);
+	} else {
+		if (completion)
+			ofi_cq_write(cq, tc->op_context, tc->flags, tc->len, tc->buf,
+				     0, tc->tag);
+
+		ofi_ep_rx_cntr_inc(&u_ep->ep);
+	}
+
+	ucx_req_release(req);
+	return ret;
+}
+
+static inline ssize_t ucx_do_inject(struct fid_ep *ep, const void *buf,
+				    size_t len, fi_addr_t dest_addr,
+				    uint64_t tag)
+{
+	struct ucx_ep *u_ep;
+	ucp_ep_h dst_ep;
+	ucs_status_ptr_t status = NULL;
+	ucs_status_t ret;
+
+	u_ep = container_of(ep, struct ucx_ep, ep.ep_fid);
+	dst_ep = UCX_GET_UCP_EP(u_ep, dest_addr);
+
+	status = ucp_tag_send_nb(dst_ep, buf, len,
+				 ucp_dt_make_contig(1),
+				 tag, ucx_send_callback_no_compl);
+
+	if (status == NULL) {
+		ofi_ep_tx_cntr_inc(&u_ep->ep);
+		return FI_SUCCESS;
+	}
+
+	if (UCS_PTR_IS_ERR(status)) {
+		FI_DBG(&ucx_prov,FI_LOG_CORE,
+		       "Send operation returns error: %s",
+		       ucs_status_string(*(ucs_status_t*)status));
+		return UCX_TRANSLATE_ERRCODE(*(ucs_status_t*) status);
+	}
+
+	while ((ret = ucp_request_check_status(status)) == UCS_INPROGRESS)
+		ucp_worker_progress(u_ep->worker);
```

Comment:
Does ucp not have support for inline data?  At least the structure of this function suggests that to me.

Reply:
There is not explicit flag in ucp to do that. It's up to the transport layer (uct) to determine if inline data is applicable. If inline send is indeed used, the send call is likely to return NULL and this function will return early (see L248-251).